src/parsers/gene_association_file.py

```
@@ -0,0 +1,117 @@
+        ]
+    
+    df = df.drop(columns = drop_columns)
+
+    # Create a unified 'subject' identifier
+    df["subject"] = df["DB"] + ":" + df["DB_Object_ID"]
+    df = df.drop(columns=["DB", "DB_Object_ID"])
+
+    # Reorder columns to place 'subject' first
+    cols = ["subject"] + [col for col in df.columns if col != "subject"]
+    df = df[cols]
+    
+    # Rename columns to match the desired schema
+    df = df.rename(columns={
+        "Qualifier": "predicate", 
+        "GO_ID": "object",
+        "DB_Reference": "publications", 
+        "With_From": "supporting_objects",
+        "Date": "annotation_date",
+        "Assigned_By": "primary_knowledge_source"})
+    
+    # Convert 'annotation_date' to a standard format
+    df["annotation_date"] = pd.to_datetime(df["annotation_date"], format='%Y%m%d').dt.strftime('%Y-%m-%d')
+    df["annotation_date"] = df["annotation_date"].astype(str)
+    
+    # Add constant metadata columns 
+    df["aggregator"] = "UniProt"
+    df["protocol_id"] = "NULL"
+    df["supporting_objects"] = df["supporting_objects"].fillna("NULL")
+    
+    # Define allowed predicates, determine negation, and clean predicate names
+    allowed_predicates = ["enables", "contributes_to", "acts_upstream_of_or_within", "involved_in",
+                          "acts_upstream_of", "acts_upstream_of_positive_effect",
+                          "acts_upstream_of_negative_effect", "acts_upstream_of_or_within_negative_effect",
+                          "acts_upstream_of_or_within_positive_effect", "located_in", "part_of", 
+                          "is_active_in", "colocalizes_with"]
+
+    df["negated"] = ~df["predicate"].isin(allowed_predicates)
+    df["predicate"] = df["predicate"].str.replace(r"^NOT\|", "", regex=True)
+    
+    # Load the external ECO mapping data
+    url = "https://raw.githubusercontent.com/evidenceontology/evidenceontology/master/gaf-eco-mapping.txt"
```

I would load this earlier in the script so that you can fail fast if the file isn't available. Also note the documentation at the top of the gar-eco-mapping file about what URL should be used.